hal/src/common/rtc.rs

```diff
@@ -300,3 +335,219 @@ impl TimerParams {
+
+    fn add(mut self, dur: Duration) -> Self {
+        self += dur;
+        self
+    }
+}
+
+impl ops::SubAssign<Duration> for Instant {
+    fn sub_assign(&mut self, dur: Duration) {
+        // NOTE see the NOTE in `<Instant as AddAssign<Duration>>::add_assign`
+        debug_assert!(dur.inner < (1 << 31));
+        self.inner = self.inner.wrapping_sub(dur.inner as i32);
+    }
+}
+
+impl ops::Sub<Duration> for Instant {
+    type Output = Self;
+
+    fn sub(mut self, dur: Duration) -> Self {
+        self -= dur;
+        self
+    }
+}
+
+impl ops::Sub<Instant> for Instant {
+    type Output = Duration;
+
+    fn sub(self, other: Instant) -> Duration {
+        self.duration_since(other)
+    }
+}
+
+impl Ord for Instant {
+    fn cmp(&self, rhs: &Self) -> Ordering {
+        self.inner.wrapping_sub(rhs.inner).cmp(&0)
+    }
+}
+
+impl PartialOrd for Instant {
+    fn partial_cmp(&self, rhs: &Self) -> Option<Ordering> {
+        Some(self.cmp(rhs))
+    }
+}
+
+/// A `Duration` type to represent a span of time.
+///
+/// # Correctness
+///
+/// This type is *not* appropriate for representing time spans in the order of,
+/// or larger than, seconds because it can hold a maximum of `(1 << 31)` "ticks"
+/// where each tick is the inverse of the CPU frequency, which usually is dozens
+/// of MHz.
+#[derive(Clone, Copy, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Duration {
+    inner: u32,
+}
+
+impl Duration {
+    /// Creates a new `Duration` from the specified number of clock cycles
+    pub fn from_cycles(cycles: u32) -> Self {
+        Duration { inner: cycles }
+    }
+
+    /// Returns the total number of clock cycles contained by this `Duration`
+    pub fn as_cycles(&self) -> u32 {
+        self.inner
+    }
+}
+
+impl TryInto<u32> for Duration {
+    type Error = Infallible;
+
+    fn try_into(self) -> Result<u32, Infallible> {
+        Ok(self.as_cycles())
+    }
+}
+
+impl ops::AddAssign for Duration {
+    fn add_assign(&mut self, dur: Duration) {
+        self.inner += dur.inner;
+    }
+}
+
+impl ops::Add<Duration> for Duration {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        Duration {
+            inner: self.inner + other.inner,
+        }
+    }
+}
+
+impl ops::SubAssign for Duration {
+    fn sub_assign(&mut self, rhs: Duration) {
+        self.inner -= rhs.inner;
+    }
+}
+
+impl ops::Sub<Duration> for Duration {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self {
+        Duration {
+            inner: self.inner - rhs.inner,
+        }
+    }
+}
+
+/// Adds the `cycles` method to the `u32` type
+pub trait U32Ext {
+    /// Converts the `u32` value into RTC counter cycles
+    fn cycles(self) -> Duration;
+}
+
+impl U32Ext for u32 {
+    fn cycles(self) -> Duration {
+        Duration { inner: self }
+    }
+}
+
+/// RtcMonotonic implements Monotonic based on the RTC peripheral for keeping
+/// track of time in RTIC
+pub struct RtcMonotonic {}
```

Comment:
Can we store `Rtc` as a non-public field here, and then provide a method to convert back from `RtcMonotonic` to `Rtc`?

Reply:
I have now changed this so that Monotonic is implemented directly on `Rtc` (which makes this obsolete)